### ISSUE TITLE
Add video and audio *encoding* support in torchcodec

### ISSUE THREAD

**Issue by ahmadsharif1** [pytorch/torchcodec#164]
The normal flow that most users have is going from video files to tensors.

However, some users do want to go in the other direction -- i.e. go from frame tensors to encoded video files. We should add that feature.

Bonus points for GPU-accelerated encoding.

**Comment by dodobyte** [comment]
I second this, there are a lot of use cases where you have to run video decoding -> GPU processing -> encoding on cloud, not necessarily encoding to disk but also streaming through WebRTC etc. 
Cloud CPU's are notoriously bad at video encoding and while the tensors are already in the GPU memory It would be great encode them there, hence have the whole pipeline handled by PyTorch.

**Comment by yassineAlouini** [comment]
So far, the best solution I have managed to come up with is using `PyNvVideoCodec` library https://docs.nvidia.com/video-technologies/pynvvideocodec/pynvc-api-prog-guide/index.html and some inspiration from this https://github.com/jxiaof/PyNvVideoCodec-example/blob/main/codec.py.

Is there something that might interest the `torchcodec ` team? I can start a small PR that does a minimal integration of a `VideoEncoder `(let's say encode an RGB tensor into a YUV420p output).

What do you think?

**Comment by NicolasHug** [comment]
Thanks for offering to help @yassineAlouini . I'm working on audio encoding at the moment, and it should be available with the next release. Video encoding will come next, but we want to be backed by FFmpeg for this, not PyNvVideoCodec. At least, for the main implementation. We'll definitely consider GPU encoding.

**Comment by yassineAlouini** [comment]
Sounds great @NicolasHug, thanks for the insights.

**Comment by cjolivier01** [comment]
Torchaudio encodes with GPU from a GPU tensor. I use it daily.

**Comment by groodt** [comment]
I stumbled up on this issue. It does appear that there is a regression in `torchaudio` caused by the adoption of `torchcodec`.

Context:
* `torchaudio==2.8.0` (torchcodec==0.7) had a capability for GPU accelerated decoding and encoding. The implementation was `libtorio` in the `torchaudio` source tree
* In the `torchaudio==2.9.0` release, `libtorio` was completely removed and all decoding and encoding was delegated to `torchcodec`
* Unfortunately, `torchcodec` does not yet have parity with `libtorio` for GPU accelerated encoding (it only has accelerated decoding)

This does make it difficult for projects that previously relied on this functionality to upgrade beyond `torchaudio==2.8.0`.

**Comment by groodt** [comment]
Looks like folks have been working behind the scenes, because GPU accelerated video encoding was implemented with [v0.10.0](https://github.com/meta-pytorch/torchcodec/releases/tag/v0.10.0)

This issue can probably be closed. GPU accelerated decoding/encoding has been implemented for Video. Audio is CPU only, but that has always been the case and audio processing doesn't typically require hardware acceleration and I don't believe there are any standard hardware acceleration approaches for audio decoding/encoding.

**Comment by NicolasHug** [comment]
Yeah, this is an old issue and it can be closed. Thanks for the ping @groodt

**Comment by NikilXYZ** [comment]
this is still not at parity - we still need an equivalent to StreamWriter, otherwise you're stuck with very short videos b/c you need to accumulate all of the frames in the buffer. I have some 20 minute or hour long videos that I am drawing bounding boxes on. But the way this lib is currently written, even a 1-2 mins might be too long for many GPUs
